README.md

```diff
@@ -51,192 +67,323 @@ and we recommend the one that begins with "https"). Let's call this the base URL
-  // You may return options synchronously or use the respond() argument to send them asynchronously
-  // See https://api.slack.com/docs/message-menus#menu_dynamic for a description of a JSON object
-  // that represents options.
-  typeaheadLookup(selection.value)
-    .then(formatMatchesAsOptions)
-    .then(respond)
-    .catch((error) => ({ options: [ { text: `Error: ${error.message}`, value: 'error' } ] }));
+slackMessages.action('create_order_dialog', (payload, respond) => {
+  // `payload` is an object that describes the interaction
+  console.log(`The user ${payload.user.name} in team ${payload.team.domain} submitted a dialog`);
+
+  // Check the values in `payload.submission` and report any possible errors
+  const errors = validateOrderSubmission(payload.submission);
+  if (errors) {
+    return errors;
+  } else {
+    setTimeout(() => {
+      // When there are no errors, after this function returns, send an acknowledgement to the user
+      respond({
+        text: `Thank you for completing an order, <@${payload.user.id}>. ` +
+              'Your order number will appear here shortly.',
+      });
+
+      // Your app does some work using information in the submission
+      orders.create(payload.submission)
+        .then((order) => {
+          // After the asynchronous work is done, call `respond()` with a message object to update the
+          // message.
+          const message = {
+            text: `Thank you for completing an order, <@${payload.user.id}>. ` +
+                  `Your order number is ${order.number}`,
+          };
+          respond(message);
+        })
+        .catch((error) => {
+          // Handle errors
+          console.error(error);
+          respond({ text: 'An error occurred while creating your order.' });
+        });
+    });
+  }
 });
+```
 
-// You can also narrow down the handler for selections coming from interactive messages versus
-// dialog suggestions
-slackMessages.options({ within: 'interactive_message' }, (selection, respond) => {
+**NOTE:** If you return a Promise which takes longer than the timeout (2.5 seconds) to complete, the
+adapter will continue to wait and the user will see an error.
 
-});
+#### Options request matching
 
-slackMessages.options({ within: 'dialog' }, (selection, respond) => {
+Use a string or RegExp as the first argument to the `.options()` method to use a `callback_id`
+constraint for the handler.
 
+```javascript
+// Run handlerFunction for any options requests from messages with a callback_id of project_menu
+slackMessages.options('project_menu', handlerFunction);
+
+// Run handlerFunction for any options requests from messages with a callback_id that match the RegExp
+slackMessages.options(/(\w+)_menu/, handlerFunction);
+
+// This function is discussed in "Responding to options requests" below
+function handlerFunction() {
+}
+```
+
+Use an object to describe other constraints, even combine multiple constraints to create more
+specific handlers.
```

Comment:
It would be good to document the breadth of possibilities of the options object. Are `callbackId` and `within` the only two keys that can be used? Are there others? What are the choices for `within` and what do each of them mean?

Reply:
I thought about this point while writing, and my view is that the matching constraints are likely not something most developers should spend too much time on, so I wanted to keep the README moving onto the "handling" portion, where the real important to understand content is. I didn't want to defer the discussion of matching constraints for after handling because its kind of essential to understanding what follows. So I was trying to keep it brief.

I'm still very much working on the reference documentation, but i'm hoping your questions are sufficiently answered there, and that its intuitive enough for you to look there for the answer. Do you think that's acceptable?

Reply:
Good point! I may be able to get away with tightening up the matching section a little more, as long as I point to the reference docs explicitly for more information.